Add a SelectorErrorCode to segment_not_present in IDT

budde25 opened the pull request:
As documented here: https://wiki.osdev.org/Exceptions#Selector_Error_Code
Similar to the `PageFaultHandlerFunc` this adds a `SegmentNotPresentHandlerFunc`

The struct does need to be a repr(transparent) u64 otherwise a LLVM error occurs

josephlr commented:
@budde25 thanks for contributing to this crate!! The definitions for this type look good, but changing the type of `segment_not_present` is a breaking change, so we would want to handle that in a separate PR (or discuss it in #262).

Could you change this PR to just declare the `SelectorErrorCode` type/methods? This would mean removing the `segment_not_present` change and `SegmentNotPresentHandlerFunc` type definitions (at least for now).

josephlr commented on src/structures/idt.rs:
```diff
@@ -918,6 +931,57 @@ bitflags! {
+
+impl SelectorErrorCode {
+    ///  When set, the exception originated externally to the processor
+    fn external(&self) -> bool {
```
This method (and the two others) should be `pub`

josephlr commented on src/structures/idt.rs:
```diff
@@ -918,6 +931,57 @@ bitflags! {
+    flags: u64,
+}
+
+impl SelectorErrorCode {
```
Could we add a `const fn` new method that took a u64 and returned either a value or an option?

budde25 replied:
Since only 16 of the bits do anything. Would it be better to take in a u16 that produce an always valid `SelectorErrorCode`? Or take a u64 that might have bits set in the reserved section and produce an Option? Otherwise we could take a u64 and wipe the most significant 48 bits.

josephlr replied:
I think we should try to have `SelectorErrorCode::new()` work a lot like [`PageFaultErrorCode::from_bits()`](https://docs.rs/x86_64/0.14.3/x86_64/structures/idt/struct.PageFaultErrorCode.html#method.from_bits). It should take in a `u64` and return a `Option<SelectorErrorCode>`, where `None` is returned if any reserved bits are set.

josephlr commented on src/structures/idt.rs:
```diff
@@ -918,6 +931,57 @@ bitflags! {
+    }
+}
+
+/// The descriptor tables.
```
This should be public, with a doc comment mentioning what the type is for

josephlr commented on src/structures/idt.rs:
```diff
@@ -918,6 +931,57 @@ bitflags! {
+// https://wiki.osdev.org/Exceptions#Selector_Error_Code
+/// Describes a segment selector error code.
+#[repr(transparent)]
+pub struct SelectorErrorCode {
```
Should we derive more traits for this object?

budde25 commented:
@josephlr 
Thanks for the code review! I will make the those changes. As for the breaking change would it be enough to just remove the change to `segment_not_present`? What would the change to `SegmentNotPresentHandlerFunc` type definitions be? or would it just be removed since it would be unused now. Thanks!

josephlr commented on src/structures/idt.rs:
```diff
@@ -918,6 +931,60 @@ bitflags! {
+
+/// The possible descriptor table values.
+///
+/// Used by the [`SelectorErrorCode`] to represent where the exception occurred.
```
```suggestion
/// Used by the [`SelectorErrorCode`] to indicate which table caused the error.
```

josephlr commented on src/structures/idt.rs:
```diff
@@ -918,6 +931,60 @@ bitflags! {
+        }
+    }
+
+    /// The index of the descriptor table.
```
```suggestion
    /// The index of the selector which caused the error.
```

josephlr commented on src/structures/idt.rs:
```diff
@@ -918,6 +931,60 @@ bitflags! {
+}
+
+impl SelectorErrorCode {
+    ///  If this flag is set, it indicates that the exception originated externally to the processor
```
```suggestion
    /// If true, indicates that the exception occurred during delivery of an event
    /// external to the program, such as an interrupt or an earlier exception.
```

josephlr commented on src/structures/idt.rs:
```diff
@@ -918,6 +931,60 @@ bitflags! {
+        self.flags.get_bit(0)
+    }
+
+    /// The descriptor table where the exception occurred.
```
```suggestion
    /// The descriptor table this error code refers to.
```

josephlr commented on src/structures/idt.rs:
```diff
@@ -634,6 +634,18 @@ pub type PageFaultHandlerFunc =
 #[derive(Copy, Clone, Debug)]
 pub struct PageFaultHandlerFunc(());
 
+/// A segment not present handler function that pushes a selector error code.
+///
+/// This type alias is only usable with the `abi_x86_interrupt` feature enabled.
+#[cfg(feature = "abi_x86_interrupt")]
+pub type SegmentNotPresentHandlerFunc =
+    extern "x86-interrupt" fn(InterruptStackFrame, error_code: SelectorErrorCode);
+
+/// This type is not usable without the `abi_x86_interrupt` feature.
+#[cfg(not(feature = "abi_x86_interrupt"))]
+#[derive(Copy, Clone, Debug)]
+pub struct SegmentNotPresentHandlerFunc(());
+
```
Remove (for now). Also when we add this back, note that the `#NP` exception isn't the only exception that uses this error code. 

`#TS`, `#NP`: `SelectorErrorCode`
`#SS`, `#GP`: `SelectorErrorCode` or zero (depending on context)
`#AC`, `#DF`: Error code always 0

`#PF`, `#CP`, `#VC`, `#SX`: Custom error code

josephlr reviewed (changes requested):
> As for the breaking change would it be enough to just remove the change to `segment_not_present`? What would the change to `SegmentNotPresentHandlerFunc` type definitions be? or would it just be removed since it would be unused now.

You're exactly correct here, I would want us to remove `SegmentNotPresentHandlerFunc` so that users don't get confused and try to use it with this crate when it won't (currently) work.

josephlr commented on src/structures/idt.rs:
```diff
@@ -918,6 +931,60 @@ bitflags! {
+    pub fn index(&self) -> u64 {
+        self.flags.get_bits(3..16)
+    }
+}
```
Would it make sense to add an `is_null()` method to `SelectorErrorCode`? The `#SS` and `#GP` sometimes return an actual selector error code, and sometimes return 0 depending on the exact reason for the error.

budde25 commented:
Alright, I have implemented the requested changes. Sorry for the delay.
I also added a `new_truncate` to the `SelectorErrorCode` as convenience method. Similar to the `from_bits_truncate` in an effort to provide a similar interface.

josephlr reviewed (approved):
LGTM, thanks for making the chanes!

> I also added a `new_truncate` to the `SelectorErrorCode` as convenience method. Similar to the `from_bits_truncate` in an effort to provide a similar interface.

Nice addition, I like it.
